File: CPP/2-Structurals/Composite/Composite.cpp
#include <iostream>
#include <vector>

// "interface" Filter class
// (lowest common denominator between single and grouped objects)
class FilterInterface
{
public:
    virtual ~FilterInterface() {}
    virtual void applyFilter() = 0;
};

class Filter: public FilterInterface
{
    // Single object class
    int _value;

public:
    Filter(int val)
    {
        _value = val;
    }

    /*virtual*/ void applyFilter()
    {
        std::cout << _value << " ";
    }
};

class FilterGroup: public FilterInterface
{
    // Group objects class
    std::vector<FilterInterface *> _children; // "container" coupled to the interface

public:
    ~FilterGroup()
    {
        for (int i = 0; i < _children.size(); i++)
        {
            delete _children[i];
        }
        _children.clear();
    }

    // "container" class coupled to the interface
    void add(FilterInterface *f)
    {
        _children.push_back(f);
    }

    /*virtual*/ void applyFilter()
    {
        for (int i = 0; i < _children.size(); i++)
        {
            // Use polymorphism to delegate to children
            _children[i]->applyFilter();
        }
    }
};

int main()
{
    // Build a single filter
    Filter *single = new Filter(42);

    // Setup a group of filters into a group
    FilterGroup *group = new FilterGroup();
    for (int i = 0; i < 4; ++i)
    {
        group->add(new Filter(i));
    }

    FilterGroup *mega_group = new FilterGroup();
    // Build a group of group of filters
    FilterGroup *group2 = new FilterGroup();
    for (int j = 5; j < 10; ++j)
    {
        group2->add(new Filter(j));
    }
    mega_group->add(group2);

    // From now on we can use the base
    // class API for single filters or groups

    // Apply the single filter
    single->applyFilter();
    std::cout << std::endl;

    // Apply the filter's group
    group->applyFilter();
    std::cout << std::endl;

    // Apply the filter's mega_group
    mega_group->applyFilter();
    std::cout << std::endl;

    delete single;
    delete group;
    delete mega_group;

    return 0;
}

//-----------------------------------------------------------
/*
Output:

42
0 1 2 3
5 6 7 8 9

*/

File: Python/2-Structurals/Composite/Composite.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod


class FilterInterface(ABC):
    """
    "interface" Filter class (lowest common denominator between single and grouped objects)
    """
    def __init__(self):
        pass

    @abstractmethod
    def apply_filter(self):
        pass


class Filter(FilterInterface):
    def __init__(self, val):
        FilterInterface.__init__(self)
        self.filter_param = val

    def apply_filter(self):
        print("Apply filter with param %d" % self.filter_param)


class FilterGroup(FilterInterface):
    def __init__(self):
        FilterInterface.__init__(self)
        self.filters = []

    def add_filter(self, f):
        self.filters.append(f)

    def apply_filter(self):
        for f in self.filters:
            f.apply_filter()


def main():
    # Setup some single and composite filters
    # ...

    # Build a single filter
    single = Filter(42)

    # Setup a group of filters into a group
    group = FilterGroup()
    for i in range(0, 4):
        group.add_filter(Filter(i))

    # Build a group of group of filters
    mega_group = FilterGroup()
    group2 = FilterGroup()
    for i in range(5, 10):
        group2.add_filter(Filter(i))
    mega_group.add_filter(group2)

    # Now we can use the base class API for single filters or groups
    single.apply_filter()
    print("")

    group.apply_filter()
    print("")

    mega_group.apply_filter()
    print("")


if __name__ == "__main__":
    main()


"""
Output:

Apply filter with param 42

Apply filter with param 0
Apply filter with param 1
Apply filter with param 2
Apply filter with param 3

Apply filter with param 5
Apply filter with param 6
Apply filter with param 7
Apply filter with param 8
Apply filter with param 9

"""
